include/crocoddyl/core/control-base.hpp

```diff
@@ -0,0 +1,150 @@
+///////////////////////////////////////////////////////////////////////////////
+// BSD 3-Clause License
+//
+// Copyright (C) 2021, LAAS-CNRS, University of Edinburgh, INRIA, University of Trento
+// Copyright note valid unless otherwise stated in individual files.
+// All rights reserved.
+///////////////////////////////////////////////////////////////////////////////
+
+#ifndef CROCODDYL_CORE_CONTROL_BASE_HPP_
+#define CROCODDYL_CORE_CONTROL_BASE_HPP_
+
+#include <vector>
+#include <string>
+#include <stdexcept>
+
+#include "crocoddyl/core/fwd.hpp"
+#include "crocoddyl/core/mathbase.hpp"
+#include "crocoddyl/core/utils/exception.hpp"
+
+namespace crocoddyl {
+
+/**
+ * @brief Abstract class for the control trajectory discretization
+ * 
+ * The control trajectory is a function of the (normalized) time.
+ * Normalized time is between 0 and 1, where 0 represents the beginning of the time step, 
+ * and 1 represents its end.
+ * The trajectory depends on the control parameters p, whose size may be larger than the
+ * size of the control inputs u.
+ *
+ */
+template <typename _Scalar>
+class ControlAbstractTpl {
+ public:
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+  typedef _Scalar Scalar;
+  typedef MathBaseTpl<Scalar> MathBase;
+  typedef typename MathBase::VectorXs VectorXs;
+  typedef typename MathBase::MatrixXs MatrixXs;
+
+  /**
+   * @brief Initialize the control dimensions
+   *
+   * @param[in] nx   Dimension of control
+   * @param[in] np   Dimension of control parameters
+   */
+  ControlAbstractTpl(const std::size_t nu, const std::size_t np);
+  ControlAbstractTpl();
+  virtual ~ControlAbstractTpl();
+
+  virtual void resize(const std::size_t nu) = 0;
+
+  /**
+   * @brief Get the value of the control at the specified time
+   *
+   * @param[in]  t      Time
+   * @param[in]  p      Control parameters
+   * @param[out] u_out  Control value
+   */
+  virtual void value(double t, const Eigen::Ref<const VectorXs>& p, Eigen::Ref<VectorXs> u_out) const = 0;
+
+  virtual VectorXs value_u(double t, const Eigen::Ref<const VectorXs>& p) const;
+
+  /**
+   * @brief Get a value of the control parameters that results in the specified control value 
+   * at the specified time
+   *
+   * @param[in]  t      Time
+   * @param[in]  u      Control value
+   * @param[out] p_out  Control parameters
+   */
+  virtual void value_inv(double t, const Eigen::Ref<const VectorXs>& u, Eigen::Ref<VectorXs> p_out) const = 0;
+
+  virtual VectorXs value_inv_p(double t, const Eigen::Ref<const VectorXs>& u) const;
+
+  /**
+   * @brief Map the specified bounds from the control space to the parameter space
+   *
+   * @param[in]  u_lb   Control lower bound
+   * @param[in]  u_ub   Control lower bound
+   * @param[out] p_lb   Control parameters lower bound
+   * @param[out] p_ub   Control parameters upper bound
+   */
+  virtual void convert_bounds(const Eigen::Ref<const VectorXs>& u_lb, const Eigen::Ref<const VectorXs>& u_ub,
```

Comment:
I don't see the reason for storing the bounds here. The bounds are already stored in the ActionModel, so we would be duplicating the information. I think the ControlParametrizationModel should simply provide a method to convert the bounds from the differential control space to the parameter space, but the values of the bounds should be stored elsewhere (as it currently is). 
Moreover, adding the bounds in the constructor seems like asking too much information to build a ControlParametrizationModel because in the part of the code where the control parametrization is created the information on the control bounds may not be available. Therefore we should at least provide a version of the constructor without bounds. This means we should have also the `set_lb` and `set_ub` methods, and we would end up replacing the current one-line instruction in `IntegratedActionModelAbstractTpl<Scalar>::init()`:
```
control_->convert_bounds(differential_->get_u_lb(), differential_->get_u_ub(), p_lb, p_ub);
```
with 4 lines of code:
```
control_->set_lb(differential_->get_u_lb());
control_->set_ub(differential_->get_u_ub());
p_lb = control_->get_lb();
p_ub = control_->get_ub();
```

Reply:
> I don't see the reason for storing the bounds here. The bounds are already stored in the ActionModel, so we would be duplicating the information. I think the ControlParametrizationModel should simply provide a method to convert the bounds from the differential control space to the parameter space, but the values of the bounds should be stored elsewhere (as it currently is).

Fair enough! Let's don't do this (as you suggested)!